### PR TITLE
Update BindGroupCommand for support of adding additional groups to a pre-existing bind and bug fixes

### PR DESCRIPTION
**Pull request by clannotguilds**
You will no longer have to unbind an entire role to add more groups. You can already add multiple groups to a single role initially, so why not be able to bind more even after a bind as already been made.

- Only checks if the role to use is not the verified or not verified role, as a group binding can be edited
- Creation of a util clamp function to clamp between 0 and 255 for the rank bindings, as before you were able to specify a rank like 1-99999 and the data file would include these numbers... could use a lot of unnecessary space!
- Prevents binding a NaN rank binding as before you could specify the syntax as 1:foo and it would save those settings, saving 'null' into your settings file, instead returns and notifies the user.
- Ensures any existing ranks are not duplicated with by creating an object lookup in which each rank is checked against to see if it already exists
- Ensures any new entries to a pre-existing binding are numerically sorted
- Adds regex anchors to prevent cases of rank strings like 's100-150f' or 'f50b' passing the existing regex match and then binding a NaN. This ensures that only Number-Number for a range or Number with no letters anywhere can pass the regex test.
- Prevents including 2 of the same rank in the rank string from being bound. A rank string of '25,50,50' will only bind 25,50.

### CODE REVIEW

**Comment by evaera** [comment]
The correct range is [0; 255] because you can bind `0` which indicates a user NOT in the group

**Comment by clannotguilds** [comment]
This has been amended now, my mistake.

File: src/commands/rover/BindGroupCommand.js [hunk 1]
```diff
@@ -66,45 +73,51 @@ class BindGroupCommand extends Command {
...
+                if (!existingRanks[i]) group.ranks.push(i)
               }
             }
-          } else if (rankNumber != null) {
-            ranks.push(rankNumber)
+          } else if (rank.match(/[\d]+/)) {
+            const rankNumber = Util.clamp(parseInt(rank), 0, 255)
+            if (!existingRanks[rank]) group.ranks.push(rankNumber)
+          } else {
+            return msg.reply(`:no_entry_sign: You have attempted to bind an invalid rank (\`${rank}\`) for group (\`${groupId}\`). Ranks should be a whole number.`)
```
Comment (evaera) [inline src/commands/rover/BindGroupCommand.js:97]: Does this break Virtual groups, like Asset, which can accept an unbounded number?

Comment (clannotguilds) [reply src/commands/rover/BindGroupCommand.js:97]: On Line 61 ranks is initialized to an empty array if there is no pre-existing bind, similar to the previous functionality of setting groups.ranks to an empty array if no rank number is specified. This PR removes various instances of setting ranks to an empty array, and instead puts it in one place.

If you mean using a command like `!bindrank <role> Asset`, without specifying a number, the resulting binding saved will still include an empty ranks array.

Comment (evaera) [reply src/commands/rover/BindGroupCommand.js:97]: No, I mean that in this code the number is calmped to 0..255. The asset virtual group needs to be able to accept a "rank" (in this case, argument) that can be larger than that.

Comment (clannotguilds) [reply src/commands/rover/BindGroupCommand.js]: This should now be solved in the latest commit.

Comment (evaera) [reply src/commands/rover/BindGroupCommand.js]: Did you test this and confirm that both regular and vritual groups work with this new behavior?

Comment (clannotguilds) [reply src/commands/rover/BindGroupCommand.js]: Yes. Running the command as `!bindrank <role> 1:266 Asset:4564646544` had produced the following in the guild data json file.
`"groupRankBindings":[{"role":"537664071999422474","groups":[{"id":"1","ranks":[255]},{"id":"Asset","ranks":[4564646544]}]}]`

Running the command a second time as `!bindrank <role> 2:266 Asset:6756765,2323` resulted in the following:
`"groupRankBindings":[{"role":"537664071999422474","groups":[{"id":"1","ranks":[255]},{"id":"Asset","ranks":[2323,6756765,4564646544]},{"id":"2","ranks":[255]}]}]`

**Review by Wolftallemo — APPROVED** [review]
From my own testing it appears to work, although at some point we should prevent users from adding duplicate ranks. Not really an issue now though since this was technically possible already.

**Comment by clannotguilds** [comment]
> From my own testing it appears to work, although at some point we should prevent users from adding duplicate ranks. Not really an issue now though since this was technically possible already.

I've updated this PR with the following:
- Adds regex anchors to prevent cases of rank strings like 's100-150f' or 'f50b' passing the existing regex match and then binding a NaN. This ensures that only Number-Number  for a range or Number with no letters anywhere can pass the regex test.
- Prevents including 2 of the same rank in the rank string from being bound. A rank string of '25,50,50' will only bind 25,50.

**Pull request closed without merging**